### DISCUSSION TITLE
Is dart debug adapter currently supported?

### DISCUSSION THREAD

**Discussion by shashanknimje** [jay-babu/mason-nvim-dap.nvim#34] [Q&A] [answered]
This is what my mason-nvim-dap.lua file looks like.

```
local dap = require("dap")

local function configureDap()
	local status_ok, dap = pcall(require, "dap")
	if not status_ok then
		print("The dap extension could not be loaded")
		return
	end
end

configureDap()

dap.set_log_level('TRACE')


require("mason-nvim-dap").setup({
	ensure_installed = {},
	-- NOTE: this is left here for future porting in case needed
	-- Whether adapters that are set up (via dap) should be automatically installed if they're not already installed.
	-- This setting has no relation with the `ensure_installed` setting.
	-- Can either be:
	--   - false: Daps are not automatically installed.
	--   - true: All adapters set up via dap are automatically installed.
	--   - { exclude: string[] }: All adapters set up via mason-nvim-dap, except the ones provided in the list, are automatically installed.
	--       Example: automatic_installation = { exclude = { "python", "delve" } }
	automatic_installation = false,
	-- Whether adapters that are installed in mason should be automatically set up in dap.
	-- Removes the need to set up dap manually.
	-- See mappings.adapters and mappings.configurations for settings.
	-- Must invoke when set to true: `require 'mason-nvim-dap'.setup_handlers()`
	-- Can either be:
	-- 	- false: Dap is not automatically configured.
	-- 	- true: Dap is automatically configured.
	-- 	- {adapters: {ADAPTER: {}, }, configurations: {ADAPTER: {}, }}. Allows overriding default configuration.	
})

require 'mason-nvim-dap'.setup_handlers {
    function(source_name)
	    -- all sources with no handler get passed here
	    -- Keep original functionality of `automatic_setup = true`
	    require('mason-nvim-dap.automatic_setup')(source_name)
    end,

    python = function(source_name)
	    dap.adapters.python = {
		    type = "executable",
		    command = "python",
		    args = {"-m","debugpy.adapter"}
	    }
	    dap.configurations.python = {
		    {
			    type = "python",
			    request = "launch",
			    name = "Launch python program",
			    -- The nvim-dap plugin populates this variable with the filename of the current buffer
			    program = "${file}",
			    -- The nvim-dap plugin populates this variable with the editor's current working directory
			    cwd = "${workspaceFolder}",
		    }
	    }
    end,

    dart = function(source_name)
	    -- Below config is for dart files that make use of flutter sdk
	    dap.adapters.dart = {
		    type = "executable",
		    command = "flutter",
		    args = {"debug_adapter"}
            }
	    dap.configurations.dart = {
		    {
			    type = "dart",
			    request = "launch",
			    name = "Launch Flutter program",
			    -- The nvim-dap plugin populates this variable with the filename of the current buffer
			    program = "${file}",
			    -- The nvim-dap plugin populates this variable with the editor's current working directory
			    cwd = "${workspaceFolder}",
			    -- This gets forwarded to the Flutter CLI tool, substitute `linux` for whatever device you wish to launch
			    -- toolArgs = {"-d", "linux"}
			    stepping_granularity = 'statement', -- 'statement' or 'line' or 'instruction'
			    -- Example of statement, line, instruction stepping granularity
			    -- Here are some examples of statements, lines of code, and instructions:
	    		    -- Statement: x = 5 (assigns the value 5 to the variable x)
	    		    -- Line of code: x = 5; y = 6 (contains two statements separated by a semicolon)
      	    		    -- Instruction: print("Hello, world!") (instructs the computer to print the string "Hello, world!" to the console)
	    	    }
	    }
    end,
}

```

I get the below error when opening neovim, no specific file.

```
mason-nvim-dap.setup_handlers: Received handler for unknown dap source name: dart.
Error detected while processing /home/shashank/.config/nvim/plugin/packer_compiled.lua:
...n-nvim-dap.nvim/lua/mason-nvim-dap/mappings/adapters.lua:85: attempt to index field 'automatic_setup' (a boolean value)
...son-nvim-dap.nvim/lua/mason-nvim-dap/automatic_setup.lua:6: loop or previous error loading module 'mason-nvim-dap.mappings.adapters'
Press ENTER or type command to continue
```

Is there a way to fix this issue? Am I doing something wrong. Please let me know. Thanks!

**Comment by jay-babu** [comment, marked as answer]
I created a PR to add support: https://github.com/jay-babu/mason-nvim-dap.nvim/pull/35. added in mason.nvim on 12/7, did not know about it: https://github.com/williamboman/mason.nvim/blob/main/lua/mason-registry/dart-debug-adapter/init.lua. Do you mind testing it?

Also, it seems you opted to not use: `automatic_setup`. May I ask why? It would remove much of the manual configuration you have to do. you can still override if you want (below). instead of `automatic_setup = true`, you pass in a table.

```lua
automatic_setup = {
  configurations = {
     dart = {
         stepping_granularity = 'statement'
     }
  },
}
```

the above would be merged with the config provided by this plugin to allow full customizability without the boilerplate.

**Comment by jay-babu** [comment]
for just dart full setup would be:

```lua
require("mason-nvim-dap").setup({
    ensure_installed = {"dart"}
    automatic_setup = {
        configurations = {
            dart = {
                stepping_granularity = "statement"
            }
        }
    }
})

require("mason-nvim-dap").setup_handlers()
```

**Comment by shashanknimje** [comment]
@jay-babu thank you so much for the super quick response brother! Super appreciate you adding support for dart also. As for your question, I am still learning a lot of things neovim and wasn't sure how automation in the context of setting up the debug adapter through mason-nvim-dap.nvim would work. And since I hadn't used it before, thought I shouldn't change too many variables at once. Any who, I am more than happy to test this out. Is the functionality already added? Do I have to update the plugin? I use Packer as my neovim manager. Let me try that and get back to you.

**Reply by jay-babu** [reply]
I did not merge it yet, can you try on the `dart` branch?

**Reply by shashanknimje** [reply]
Oh okay. So in that case, if you don't mind me asking, how do I install the dart branch version if at all using packer nvim. If not do I have to clone the repo and do it manually somehow (might take longer as I haven't done that before).

**Reply by jay-babu** [reply]
https://github.com/wbthomason/packer.nvim example:

```
  -- Use specific branch, dependency and run lua file after load
  use {
    'glepnir/galaxyline.nvim', branch = 'main', config = function() require'statusline' end,
    requires = {'kyazdani42/nvim-web-devicons'}
  }
```

simply add `branch = "dart"` and run `:PackerSync`

**Reply by shashanknimje** [reply]
Thank you! Let me give it a shot in that case.

**Reply by shashanknimje** [reply]
Okay, I think it is working. But I don't want to confirm unless I am really sure. So the dart debug adapter in question supports debugging both regular dart programs as well as the dart programs using the flutter framework. For the latter below is a part of the config for nvim-dap plugin.

dap.adapters.dart = {
	type = "executable",
    	command = "flutter",
    	args = {"debug_adapter"}
}

So for flutter sdk the command would be "flutter" and for regular dart programs it would simply be "dart". How can I specify this in mason-nvim-dap.lua config file? Please let me know. Thanks!

**Reply by jay-babu** [reply]
https://github.com/mfussenegger/nvim-dap/wiki/Debug-Adapter-installation#Dart

I followed this. It should support both, I think

**Reply by jay-babu** [reply]
https://github.com/jay-babu/mason-nvim-dap.nvim/pull/35/files#diff-699dd04d8e33cb71e19790bc0456d7294def87fa40c1b976257ef7b4bad0a79fR87-R88

**Reply by shashanknimje** [reply]
> https://github.com/mfussenegger/nvim-dap/wiki/Debug-Adapter-installation#Dart
> 
> I followed this. It should support both, I think

Looks like you're right. Mason installs the vs-code plugin dart-code. I used it before google made the dart / flutter debug adapter available as part of the sdk if I'm not wrong. That is why my config above looks different from the one linked here.

So as per the below example.

  dap.adapters.dart = {
    type = "executable",
    command = "node",
    args = {"<path-to-Dart-Code>/out/dist/debug.js", "flutter"}
  }
  dap.configurations.dart = {
    {
      type = "dart",
      request = "launch",
      name = "Launch flutter",
      dartSdkPath = os.getenv('HOME').."/flutter/bin/cache/dart-sdk/",
      flutterSdkPath = os.getenv('HOME').."/flutter",
      program = "${workspaceFolder}/lib/main.dart",
      cwd = "${workspaceFolder}",
    }
  }

Correct me if I'm wrong, but I think I will have to override dartSdkPath and flutterSdkPath inside dap.configurations.dart, and as for the arguments for the node command inside dap.adapters.dart definition, do I have to do something about that? I see that you have pointed me to the below link.

https://github.com/jay-babu/mason-nvim-dap.nvim/pull/35/files#diff-699dd04d8e33cb71e19790bc0456d7294def87fa40c1b976257ef7b4bad0a79fR87-R88

Where you have added the below code.

M.dart = {
	type = 'executable',
	command = 'dart-debug-adapter',
	args = { 'flutter' },
}

Does it automatically pick the first argument below? i.e., the path to debug.js.
args = {"<path-to-Dart-Code>/out/dist/debug.js", "flutter"}

**Reply by jay-babu** [reply]
> Can I think I will have to override dartSdkPath and flutterSdkPath inside dap.configurations.dart, and as for the arguments for the node command inside dap.adapters.dart definition, do I have to do something about that? I see that you have pointed me to the below link.

if you wish to override that is fine. you can override using https://github.com/jay-babu/mason-nvim-dap.nvim#default-configuration and following the example from earlier.

> Does it automatically pick the first argument below? i.e., the path to debug.js

yes

**Reply by jay-babu** [reply]
Seems like it is working. I will merge the PR

**Reply by shashanknimje** [reply]
> > Can I think I will have to override dartSdkPath and flutterSdkPath inside dap.configurations.dart, and as for the arguments for the node command inside dap.adapters.dart definition, do I have to do something about that? I see that you have pointed me to the below link.
> 
> if you wish to override that is fine. you can override using https://github.com/jay-babu/mason-nvim-dap.nvim#default-configuration and following the example from earlier.
> 
> > Does it automatically pick the first argument below? i.e., the path to debug.js
> 
> yes

Thanks much for you guidance on this. Do I have to override the below code somehow?

dap.adapters.dart = {
type = "executable",
command = "node",
args = {"/out/dist/debug.js", "flutter"}
}

Meaning, do I have to pass "flutter" as an argument to run the debug adapter in flutter sdk mode? or does that run by default also? And if so, would that cover regular dart programs (non flutter sdk) also? Sorry to bug you with all these questions.

**Reply by jay-babu** [reply]
if you use automatic_setup then this will be the default

https://github.com/jay-babu/mason-nvim-dap.nvim/blob/a775db8ac7c468fb05fcf67069961dba0d7feb56/lua/mason-nvim-dap/mappings/adapters.lua#L85-L89

https://github.com/jay-babu/mason-nvim-dap.nvim/blob/a775db8ac7c468fb05fcf67069961dba0d7feb56/lua/mason-nvim-dap/mappings/configurations.lua#L217-L227

if you want to override it, then

```lua
require("mason-nvim-dap").setup({
    ensure_installed = {"dart"}
    automatic_setup = {
        adapter = {
            -- whatever you want
        },
        configurations = {
            dart = {
                -- whatever you want to override
            }
        }
    }
})

require("mason-nvim-dap").setup_handlers()
```

**Reply by shashanknimje** [reply]
Thanks much for the detailed response @jay-babu. So I have some not so good news. 

Turns out I had a flutter-tool-nvim plugin that was actually the one handling all the dart debug adapter work in conjunction with nvim-dap. So I removed it in order to make sure that mason-nvim-dap is able to do the same. And I also removed the nvim-dap.lua config in order to make sure the mason-nvim-dap.lua is the one that gets picked up. 

For some reason it is unable to launch the debug adapter when I try to do so in a flutter file, and it gives me the below error.

```
No configuration found for 'dart'. You need to add configs to 'dap.configurations.dart' (See ':h dap-configuration')
```

Here are the relevant plugins I have installed at the moment.
```
use ({"mfussenegger/nvim-dap"}) -- left out config = get_config("nvim-dap") for mason-nvim-dap plugin

use {"rcarriga/nvim-dap-ui", config = get_config("nvim-dap-ui"), requires = {"mfussenegger/nvim-dap"}}

use {"williamboman/mason.nvim", config = get_config("mason-nvim")}

use ({"jay-babu/mason-nvim-dap.nvim", config = get_config("mason-nvim-dap")})
```

Here is my current mason-nvim-dap.lua config file. I do realize that I don't need to override 'args' inside the dart adapter settings, but when I tried debugging using ":DapContinue" I got the above mentioned error message, and hence I tried to make sure everything is setup correctly, you know just in case.
```
local dap = require("dap")
local home = os.getenv("HOME")

local function configureDap()
	local status_ok, dap = pcall(require, "dap")
	if not status_ok then
		print("The dap extension could not be loaded")
		return
	end
end

configureDap()

dap.set_log_level('TRACE')

require("mason-nvim-dap").setup({

	ensure_installed = {"dart", "python"},

	automatic_installation = true, -- Installs the debug adapter(s) via Mason that aren't already installed, and are included in ensure_installed above

	automatic_setup = {
		adapters = {
			dart = {
				args = {
					"flutter"
				}
			}
		},
		configurations = {
			dart = {
				stepping_granularity = "statement",
				dartSdkPath = home .. "/development/flutter/bin/cache/dart-sdk",
				flutterSdkPath = home .. "/development/flutter"
			}
		}
	},
})

require 'mason-nvim-dap'.setup_handlers()
```

Also I tried to debug a python file using ":DapContinue" and that worked just fine. So I know it is just the dart file that it's unable to debug. Please let me know what I can do to fix this. Or if you want me to do any further testing. Thanks much for your help on this.

**Reply by jay-babu** [reply]
try this.

`vim.tbl_deep_extend` overwrites lists when trying to merge 2 lists.

```lua
local dap = require("dap")
local home = os.getenv("HOME")
local mason_packages = "/.local/share/nvim/mason/packages"

local function configureDap()
	local status_ok, dap = pcall(require, "dap")
	if not status_ok then
		print("The dap extension could not be loaded")
		return
	end
end

configureDap()

dap.set_log_level('TRACE')

require("mason-nvim-dap").setup({

	ensure_installed = {"dart", "python"},

	automatic_installation = true, -- Installs the debug adapter(s) via Mason that aren't already installed, and are included in ensure_installed above

	automatic_setup = {
		configurations = {
			dart = {
	                          {
		                          type = 'dart',
		                          request = 'launch',
		                          name = 'Launch flutter',
				          stepping_granularity = "statement",
				          dartSdkPath = home .. "/development/flutter/bin/cache/dart-sdk",
				          flutterSdkPath = home .. "/development/flutter"
		                          program = '${workspaceFolder}/lib/main.dart',
		                          cwd = '${workspaceFolder}',
	                          },
                          },
		},
	},
})

require 'mason-nvim-dap'.setup_handlers()
```

**Reply by shashanknimje** [reply]
Thanks @jay-babu, that did not resolve the problem either. Here's what my config looks like now.
```
local dap = require("dap")
local home = os.getenv("HOME")

local function configureDap()
	local status_ok, dap = pcall(require, "dap")
	if not status_ok then
		print("The dap extension could not be loaded")
		return
	end
end

configureDap()

dap.set_log_level('TRACE')

require("mason-nvim-dap").setup({

	ensure_installed = {"dart", "python"},

	automatic_installation = true, -- Installs the debug adapter(s) via Mason that aren't already installed, and are included in ensure_installed above

	automatic_setup = {

		--[==[
		adapters = {
			dart = {
				type = 'executable',
				command = "flutter", -- needed only when not using the Mason provided (vs-code debu.js) adapter
				args = {"debug_adapter"} -- needed only when not using the Mason provided (vs-code debu.js) adapter
			}
		},
		--]==]

		configurations = {
			dart = {
				type = 'dart',
				request = 'launch',
				name = 'Launch flutter',
				stepping_granularity = "statement",
				dartSdkPath = home .. "/development/flutter/bin/cache/dart-sdk", -- needed only for vs-code dart-code (debug.js) plugin adapter
				flutterSdkPath = home .. "/development/flutter", -- needed only for vs-code dart-code (debug.js) plugin adapter
				program = "${workspaceFolder}/lib/main.dart", -- needed only for vs-code dart-code (debug.js) plugin adapter
				--program = "${file}", -- needed only when not using the Mason provided (vs-code debu.js) adapter
				cwd = '${workspaceFolder}'
			}
		}
	},
})

require 'mason-nvim-dap'.setup_handlers()
```

Is it working for you?
